Convert natgateways service to SDKv2

mboersma opened the pull request:
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Converts the natgateways service to SDKv2 and a new `asyncpoller` framework for long-running operations.

**Which issue(s) this PR fixes**:

Fixes #3442
Fixes #3441

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
NONE
```

mboersma commented:
I rebased this (temporarily) on #3843 and made the necessary changes to use that apporach for authentication.

mboersma commented:
This is ready for review if anyone has time.

But I'd like to get #3857 full working as well before we merge this, so we have high confidence in the `asyncpoller` framework.

mboersma commented on azure/services/asyncpoller/interfaces.go:
```diff
@@ -0,0 +1,67 @@
+	azure.AsyncStatusUpdater
+}
+
+// FutureHandler is a client that can check on the progress of a poller.
+type FutureHandler interface {
+	// IsDone returns true if the operation is complete.
+	IsDone(ctx context.Context, poller interface{}) (isDone bool, err error)
+	// Result returns the result of the operation.
+	Result(ctx context.Context, poller interface{}) (result interface{}, err error)
+}
```
It appears we can. I started with a 1:1 port of `async` and refactored away the need for this but must have kept the code.

mboersma commented:
/hold

For more reviews, testing, and eventual squash.

mboersma commented on azure/services/natgateways/client.go:
```diff
@@ -18,139 +18,145 @@ package natgateways
 	defer cancel()
 
-	err = createFuture.WaitForCompletionRef(ctx, ac.natgateways.Client)
+	result, err = poller.PollUntilDone(ctx, &runtime.PollUntilDoneOptions{})
```
Done--set to a constant of 1.

mboersma commented on azure/services/asyncpoller/asyncpoller.go:
```diff
@@ -0,0 +1,190 @@
+
+// getRequeueAfterFromPoller returns the max between the `RETRY-AFTER` header and the default requeue time.
+// This ensures we respect the retry-after header if it is set and avoid retrying too often during an API throttling event.
+// TODO: there doesn't seem to be a replacement for sdkFuture.GetPollingDelay() in the new poller.
```
I don't see a reasonable way to get at the response from that context. I removed this function and just use the default time explicitly.

mboersma replied:
Updated: I left a function in place so we could document that using the "retry-after" header would be a better approach.

mboersma commented on azure/services/asyncpoller/asyncpoller_test.go:
```diff
@@ -0,0 +1,221 @@
+					s.GetLongRunningOperationState(resourceName, serviceName, infrav1.PutFuture).Return(validPutFuture),
+					c.Get(gomockinternal.AContext(), gomock.AssignableToTypeOf(azureResourceGetterType)).Return(fakeResource, nil),
+					r.Parameters(gomockinternal.AContext(), fakeResource).Return(fakeParameters, nil),
+					c.CreateOrUpdateAsync(gomockinternal.AContext(), gomock.AssignableToTypeOf(azureResourceGetterType), resumeToken, gomock.Any()).Return(fakeResource, fakePoller[MockCreator](), nil),
```
Added, good call.

mboersma commented on azure/services/asyncpoller/asyncpoller.go:
```diff
@@ -156,13 +160,6 @@ func (s *Service[C, D]) DeleteResource(ctx context.Context, spec azure.ResourceS
 
-// getRequeueAfterFromPoller returns the max between the `RETRY-AFTER` header and the default requeue time.
-// This ensures we respect the retry-after header if it is set and avoid retrying too often during an API throttling event.
-// TODO: there doesn't seem to be a replacement for sdkFuture.GetPollingDelay() in the new poller.
```
I left a function in place to document that using the "retry-after" header would be a better approach.

nojnhuh commented on azure/services/natgateways/spec.go:
```diff
@@ -91,8 +91,8 @@ func (s *NatGatewaySpec) Parameters(ctx context.Context, existing interface{}) (
 func hasPublicIP(natGateway armnetwork.NatGateway, publicIPName string) bool {
-	// We must have a non-nil, non-"empty" PublicIPAddresses
-	if !(natGateway.Properties.PublicIPAddresses != nil && len(natGateway.Properties.PublicIPAddresses) > 0) {
+	// PublicIPAddresses must not be nil or empty.
+	if len(natGateway.Properties.PublicIPAddresses) == 0 {
 		return false
 	}
```
I suppose we could get rid of this entirely since an empty slice will basically skip over the for loop and return false anyway.

nojnhuh reviewed (commented):
/lgtm

mboersma commented:
I squashed the commits and added some additional unit test paths through `CreateOrUpdateAsync`. 

The hope is that the second commit here https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3858/commits/517cfe2773ea7a0aef2ea9b875e832f96ceaf866 will serve as a reference for others to convert services to SDKv2 / `asyncpoller`. Please keep that in mind and suggest improvements, additional comments, etc.

nojnhuh commented:
LGTM pending review from at least one other person.

willie-yao reviewed (approved):
This looks great! Thanks for your hard work on this @mboersma 
/lgtm

mboersma commented:
/retest

Looks like a provisioning flake in ci-entrypoint: timeout `# Wait for the kubeconfig to become available.`

nojnhuh commented:
/lgtm
/approve

mboersma commented:
/hold cancel
